
# 🧩 Titanic Dataset Exploration with pandas

🎯 **Goal:**  
In this notebook, you'll explore the Titanic dataset using **pandas**.  
You'll practice the most common pandas functions for data inspection, selection, filtering, cleaning, and analysis.

For each function in the list below:
1. Explain what it does (in your own words, in a Markdown cell).
2. Give at least **two examples** using the Titanic dataset.
3. Add a short comment about the output or why it’s useful.


In [2]:
import pandas as pd

# Load Titanic dataset
# (Make sure titanic.csv is in the same folder as this notebook)
df = pd.read_csv("titanic_dataset.csv")

# Show first few rows
df.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S


## 🧠 Step 1: Inspecting the Data

Functions to explore:
- df.head()
- df.tail()
- df.info()
- df.describe()
- df.shape
- df.columns


## df.head()

### 1. Descripción:  
Es una función que por defecto devuelve las primeras 5 filas del dataframe.   
Si se pasa un int por parámetro, devuelve las primeras n filas.   
Si se pasa un int negativo por parámetro, devolverá todas las filas excepto las n últimas.  

<br>

### 2. Ejemplos:

In [38]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Fare_per_Age  
0      0         A/5 21171   7.2500   NaN        S      0.329545  
1      0          PC 17599  71.2833   C85        C      1.875876  
2      0  STON/O2. 3101282   7.9250   NaN        S      0.304808

In [41]:
df.head(-91)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
795          796         0       2   
796          797         1       1   
797          798         1       3   
798          799         0       3   
799          800         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
795                                 Otter, Mr. Richard    male  39.0      0   
796                        Leader, Dr. Alice (Farnham)  female  49.0      0   
797                                   Osman, Mrs. Mara  female  31.0      0   
798                       Ibrahim Shawah, Mr. Yousseff    male  30.0      0   
799  Van Impe, Mrs. Jean Baptiste (Rosalie Paula Go...  female  30.0      1   

     Parch            Ticket     Fare Cabin Embarked  Fare_per_Age  
0        0         A/5 21171   7.2500   NaN        S      0.329545  
1        0          PC 17599  71.2833   C85        C      1.875876  
2        0  STON/O2. 3101282   7.9250   NaN        S      0.304808  
3        0            113803  53.1000  C123        S      1.517143  
4        0            373450   8.0500   NaN        S      0.230000  
..     ...               ...      ...   ...      ...           ...  
795      0             28213  13.0000   NaN        S      0.333333  
796      0             17465  25.9292   D17        S      0.529167  
797      0            349244   8.6833   NaN        S      0.280106  
798      0              2685   7.2292   NaN        C      0.240973  
799      1            345773  24.1500   NaN        S      0.805000  

[800 rows x 13 columns]

### 3. Utilidad

Permite de manera rápida ver la estructura de un DataFrame.  

Si se ordenan los datos, nos permite crear nuevos DataFrames filtrando por los n primeros registros.  

Por ejemplo, si los datos estuviesen ordenados por edad:

<br>

**50 pasajeros más jóvenes** 

```python
df_50_pasajeros_mas_jovenes = df. head(50)
``` 
<br>

**Todos los pasajeros salvo los 50 más ancianos**

```python 
df_sin_50_mas_ancianos = df.head(-50)
```

---


## df.tail()

### 1. Descripción:  
Es una función que por defecto devuelve las primeras 5 últimas filas del dataframe.   
Si se pasa un int por parámetro, devuelve las últimas n filas.   
Si se pasa un int negativo por parámetro, devolverá todas las filas excepto las n primeras.  

<br>

### 2. Ejemplos:

In [35]:
df.tail(3)

PassengerId  Survived  Pclass                                      Name  \
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  \
888  female  28.0      1      2  W./C. 6607  23.45   NaN        S   
889    male  26.0      0      0      111369  30.00  C148        C   
890    male  32.0      0      0      370376   7.75   NaN        Q   

     Fare_per_Age  
888      0.837500  
889      1.153846  
890      0.242188

In [44]:
df.tail(-100)

PassengerId  Survived  Pclass                                      Name  \
100          101         0       3                   Petranec, Miss. Matilda   
101          102         0       3          Petroff, Mr. Pastcho ("Pentcho")   
102          103         0       1                 White, Mr. Richard Frasar   
103          104         0       3                Johansson, Mr. Gustaf Joel   
104          105         0       3            Gustafsson, Mr. Anders Vilhelm   
..           ...       ...     ...                                       ...   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  \
100  female  28.0      0      0      349245   7.8958   NaN        S   
101    male  28.0      0      0      349215   7.8958   NaN        S   
102    male  21.0      0      1       35281  77.2875   D26        S   
103    male  33.0      0      0        7540   8.6542   NaN        S   
104    male  37.0      2      0     3101276   7.9250   NaN        S   
..      ...   ...    ...    ...         ...      ...   ...      ...   
886    male  27.0      0      0      211536  13.0000   NaN        S   
887  female  19.0      0      0      112053  30.0000   B42        S   
888  female  28.0      1      2  W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0      111369  30.0000  C148        C   
890    male  32.0      0      0      370376   7.7500   NaN        Q   

     Fare_per_Age  
100      0.281993  
101      0.281993  
102      3.680357  
103      0.262248  
104      0.214189  
..            ...  
886      0.481481  
887      1.578947  
888      0.837500  
889      1.153846  
890      0.242188  

[791 rows x 13 columns]

### 3. Utilidad

Permite de manera rápida ver la estructura de un DataFrame.  

Si se ordenan los datos, nos permite crear nuevos DataFrames filtrando por los n últimos registros.  

Por ejemplo, si los datos estuviesen ordenados por edad:

<br>

**50 pasajeros más ancianos** 

```python
df_50_pasajeros_mas_ancianos = df.tail(50)
``` 
<br>

**Todos los pasajeros salvo los 50 más jóvenes**

```python 
df_sin_50_mas_jovenes = df.tail(-50)
```

---

## df.info()

### 1. Descripción:  
Imprime un resumen conciso de un DataFrame. Incluyendo el índices, columnas, dtypes, cantidad de no nulos y uso de memoria.

<br>

### 2. Ejemplos:

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   891 non-null    int64  
 1   Survived      891 non-null    int64  
 2   Pclass        891 non-null    int64  
 3   Name          891 non-null    object 
 4   Sex           891 non-null    object 
 5   Age           891 non-null    float64
 6   SibSp         891 non-null    int64  
 7   Parch         891 non-null    int64  
 8   Ticket        891 non-null    object 
 9   Fare          891 non-null    float64
 10  Cabin         204 non-null    object 
 11  Embarked      889 non-null    object 
 12  Fare_per_Age  891 non-null    float64
dtypes: float64(3), int64(5), object(5)
memory usage: 90.6+ KB


### 3. Utilidad

Permite ver de manera rápida la estructura de un DataFrame.

---

## df.describe()

### 1. Descripción:  
Genera estadísticas descriptivas del DataFrame.

<br>

### 2. Ejemplos:

In [55]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.361582    0.523008   
std     257.353842    0.486592    0.836071   13.019697    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  Fare_per_Age  
count  891.000000  891.000000    891.000000  
mean     0.381594   32.204208      2.073904  
std      0.806057   49.693429      7.309062  
min      0.000000    0.000000      0.000000  
25%      0.000000    7.910400      0.299519  
50%      0.000000   14.454200      0.535341  
75%      0.000000   31.000000      1.417335  
max      6.000000  512.329200    164.728261

In [56]:
df.describe(include='object')

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

### 3. Utilidad

La función .describe() de pandas permite obtener un resumen estadístico rápido de las columnas numéricas de un DataFrame.  
Identifica valores extremos, medias, medianas y percentiles.  
También puede aplicarse a columnas no numéricas con:
```python
df.describe(include='object')
```

Este incluirá:   
count (cantidad de valores)  
unique (número de valores únicos),   
top (el valor más frecuente)  
freq (la frecuencia del valor más frecuente)  

---

## df.shape

### 1. Descripción:  
Devuelve una tupla que representa las dimesiones del DataFrame

<br>

### 2. Ejemplos:

In [59]:
df.shape

(891, 13)

### 3. Utilidad

Para el dataset de los pasajeros del Titanic indicaría que se trata de una tabla de 13 columnas por 891 filas.

---

## df.columns()

### 1. Descripción:  
Devuelve todas las etiquetas de las columnas del DataFrame.

<br>

### 2. Ejemplos:

In [61]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Fare_per_Age'],
      dtype='object')

### 3.Utilidad

Podría ser util si obtenemos la lista de etiquetas para iterarla en algún bucle.

---


## 🔍 Step 2: Selecting Data

Functions to explore:
- df["column"]
- df[["col1", "col2"]]
- df.loc[]
- df.iloc[]


## df["column"]

### 1. Descripción:  
Devuelve los registros de la columna especificada en formato **Serie**.  


### 2. Ejemplos:


In [65]:
df["Age"].head()



0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

### 3.Utilidad

Para ver solo aquella o aquellas columnas en las que queramos trabajar.

---

## df["col1", "col2"]

### 1. Descripción:  
Devuelve los registros de las columnas especificadas en formato **DataFrame**.  


### 2. Ejemplos:


In [68]:
df[["Sex", "Age", "Survived"]].head()

Sex   Age  Survived
0    male  22.0         0
1  female  38.0         1
2  female  26.0         1
3  female  35.0         1
4    male  35.0         0

### 3.Utilidad

Para crear un sub-DataFrame con varias columnas para análisis o visualización.

---

## df.loc[]

### 1. Descripción:    

Accede a un grupo de filas y columnas a través de sus etiquetas.


### 2. Ejemplos:

In [ ]:
df.loc[(df['Age'] < 18) & (df['Survived'] == 0), ['Name', 'Age']]

Name   Age
7             Palsson, Master. Gosta Leonard   2.0
14      Vestrom, Miss. Hulda Amanda Adolfina  14.0
16                      Rice, Master. Eugene   2.0
24             Palsson, Miss. Torborg Danira   8.0
50                Panula, Master. Juha Niilo   7.0
59        Goodwin, Master. William Frederick  11.0
63                     Skoog, Master. Harald   4.0
71                Goodwin, Miss. Lillian Amy  16.0
86                    Ford, Mr. William Neal  16.0
111                     Zabour, Miss. Hileni  14.5
114                    Attalah, Miss. Malake  17.0
119        Andersson, Miss. Ellis Anna Maria   2.0
138                      Osen, Mr. Olaf Elon  16.0
147         Ford, Miss. Robina Maggie "Ruby"   9.0
163                          Calic, Mr. Jovo  17.0
164             Panula, Master. Eino Viljami   1.0
171                     Rice, Master. Arthur   4.0
182    Asplund, Master. Clarence Gustaf Hugo   9.0
205               Strom, Miss. Telma Matilda   2.0
266                Panula, Mr. Ernesti Arvid  16.0
278                       Rice, Master. Eric   7.0
282                de Pelsmaeker, Mr. Alfons  16.0
297             Allison, Miss. Helen Loraine   2.0
333          Vander Planke, Mr. Leo Edmondus  16.0
352                       Elias, Mr. Tannous  15.0
374               Palsson, Miss. Stina Viola   3.0
386          Goodwin, Master. Sidney Leonard   1.0
419                Van Impe, Miss. Catharina  10.0
433               Kallio, Mr. Nikolai Erland  17.0
480           Goodwin, Master. Harold Victor   9.0
500                         Calic, Mr. Petar  17.0
532                     Elias, Mr. Joseph Jr  17.0
541     Andersson, Miss. Ingeborg Constanzia   9.0
542        Andersson, Miss. Sigrid Elisabeth  11.0
574             Rush, Mr. Alfred George John  16.0
634                       Skoog, Miss. Mabel   9.0
642            Skoog, Miss. Margit Elizabeth   2.0
683              Goodwin, Mr. Charles Edward  14.0
686                 Panula, Mr. Jaako Arnold  14.0
721                Jensen, Mr. Svend Lauritz  17.0
731                 Hassan, Mr. Houssein G N  11.0
746              Abbott, Mr. Rossmore Edward  16.0
764                   Eklund, Mr. Hans Linus  16.0
787                Rice, Master. George Hugh   8.0
791                      Gaskell, Mr. Alfred  16.0
813       Andersson, Miss. Ebba Iris Alfrida   6.0
819             Skoog, Master. Karl Thorsten  10.0
824             Panula, Master. Urho Abraham   2.0
841                 Mudd, Mr. Thomas Charles  16.0
844                      Culumovic, Mr. Jeso  17.0
850  Andersson, Master. Sigvard Harald Elias   4.0
852                  Boulos, Miss. Nourelain   9.0

In [12]:
df.loc[df['Sex']=='female','Name']

1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
8      Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                    Nasser, Mrs. Nicholas (Adele Achem)
                             ...                        
880         Shelley, Mrs. William (Imanita Parrish Hall)
882                         Dahlberg, Miss. Gerda Ulrika
885                 Rice, Mrs. William (Margaret Norton)
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
Name: Name, Length: 314, dtype: object

### 3.Utilidad

En el primer ejemplo calculo un dataframe con los pasajeros fallecidos menores de edad, en las filas paso las condiciones y en las columnas imprimo nombre y edad, al imprimir varias columnas me devuelve un DataFrame.

En el segundo ejemplo imprimo los nombres de todas las mujeres del barco, al imprimir solo una columna me devuelve una Serie.

---

## df.iloc[]

### 1. Descripción:  



### 2. Ejemplos:


### 3.Utilidad


---


## 🔎 Step 3: Filtering Rows

Functions to explore:
- df[df["Age"] > 30]
- df.query("Sex == 'female' and Survived == 1")


In [24]:

# Example: Filtering data
df[df["Age"] > 50].head()
df.query("Sex == 'female' and Survived == 1").head()


PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C


## 🧹 Step 4: Handling Missing Data

Functions to explore:
- df.isna()
- df.isna().sum()
- df.dropna()
- df.fillna()


In [25]:

# Example: Check missing values
df.isna().sum()

# Fill missing ages with median
df["Age"] = df["Age"].fillna(df["Age"].median())
df["Age"].head()


0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64


## 📊 Step 5: Grouping and Aggregating

Functions to explore:
- df.groupby("Sex")["Survived"].mean()
- df["Fare"].mean()
- df["Age"].median()


In [26]:

# Example: Aggregation
df.groupby("Sex")["Survived"].mean()
df.groupby("Pclass")["Fare"].mean()


Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64


## 📈 Step 6: Sorting and Counting

Functions to explore:
- df.sort_values("Age")
- df["Sex"].unique()
- df["Pclass"].value_counts()


In [27]:

# Example: Sorting and counting
df.sort_values("Age").head()
df["Pclass"].value_counts()


Pclass
3    491
1    216
2    184
Name: count, dtype: int64


## ⚙️ Step 7: Creating or Modifying Columns

Functions to explore:
- df.assign()
- df.apply()
- df["new_col"] = ...
- pd.concat()
- pd.merge()


In [28]:

# Example: Create new column
df["Fare_per_Age"] = df["Fare"] / df["Age"]
df[["Age", "Fare", "Fare_per_Age"]].head()


Age     Fare  Fare_per_Age
0  22.0   7.2500      0.329545
1  38.0  71.2833      1.875876
2  26.0   7.9250      0.304808
3  35.0  53.1000      1.517143
4  35.0   8.0500      0.230000


## 💾 Step 8: Exporting Data

Function to explore:
- df.to_csv("output.csv", index=False)


In [29]:

# Example: Save cleaned data
df.to_csv("titanic_cleaned.csv", index=False)



## 🧩 Step 9: Summary

Reflect on what you learned:
- Which functions were most useful?
- What insights did you gain from the Titanic dataset?
